bin/faucet/src/server/pow.rs

```diff
@@ -34,119 +30,90 @@ const ACTIVE_REQUESTS_TO_INCREASE_DIFFICULTY: usize = REQUESTS_QUEUE_SIZE / MAX_
-    ///
-    /// The timestamp is valid if it is within `SERVER_TIMESTAMP_TOLERANCE_SECONDS` seconds of the
-    /// current time.
-    pub fn check_server_timestamp(&self) -> Result<(), InvalidRequest> {
-        let server_timestamp = SystemTime::now()
-            .duration_since(UNIX_EPOCH)
-            .expect("Time went backwards")
-            .as_secs();
-
-        if (server_timestamp - self.server_timestamp) > SERVER_TIMESTAMP_TOLERANCE_SECONDS {
-            return Err(InvalidRequest::ExpiredServerTimestamp(
-                self.server_timestamp,
-                server_timestamp,
-            ));
+    /// Decodes the challenge and verifies that the signature part of the challenge is valid
+    /// in the context of the specified salt.
+    pub fn decode(value: &str, salt: [u8; 32]) -> Result<Self, InvalidRequest> {
+        // Add the "0x" prefix to the hex string, otherwise the hex_to_bytes function will fail.
+        let value = "0x".to_string() + value;
+        // Parse the hex-encoded challenge string
+        let bytes = hex_to_bytes::<48>(&value).map_err(|_| InvalidRequest::MissingPowParameters)?;
+
+        if bytes.len() != 48 {
+            // 8 + 8 + 32 bytes
+            return Err(InvalidRequest::MissingPowParameters);
         }
 
-        Ok(())
-    }
+        // SAFETY: Length of 48 is enforced above.
+        let difficulty = u64::from_le_bytes(bytes[0..8].try_into().unwrap()) as usize;
+        let timestamp = u64::from_le_bytes(bytes[8..16].try_into().unwrap());
+        let signature: [u8; 32] = bytes[16..48].try_into().unwrap();
 
-    /// Check a `PoW` solution.
-    ///
-    /// * `challenge_cache` - The challenge cache to be used to validate the challenge.
-    ///
-    /// The solution is valid if the hash of the seed and the solution has at least `DIFFICULTY`
-    /// leading zeros.
-    pub fn check_pow_solution(
-        &self,
-        challenge_cache: &ChallengeCache,
-    ) -> Result<(), InvalidRequest> {
-        let mut challenges =
-            challenge_cache.challenges.lock().expect("PoW challenge cache lock poisoned");
+        let challenge = Self::new(difficulty, timestamp, signature);
 
-        if challenges.get(&self.pow_seed).is_some() {
-            return Err(InvalidRequest::ChallengeAlreadyUsed);
+        // Verify the signature
+        let expected_signature = Self::compute_signature(salt, difficulty, timestamp);
+        if signature != expected_signature {
+            return Err(InvalidRequest::ServerSignaturesDoNotMatch);
         }
 
-        // Then check the PoW solution
+        Ok(challenge)
+    }
+
+    /// Encodes the challenge into a hex string.
+    pub fn encode(&self) -> String {
+        let mut bytes = Vec::with_capacity(48);
+        bytes.extend_from_slice(&(self.difficulty as u64).to_le_bytes());
+        bytes.extend_from_slice(&self.timestamp.to_le_bytes());
+        bytes.extend_from_slice(&self.signature);
+        bytes.to_hex()
+    }
```

Comment:
Related to one of the above comments: the hex string we output here should probably be prefixed with `0x`.

Reply:
We could go either way, but it's pretty easy to output the string here with `0x` prefix (using `.to_hex_with_prefix()` as you mentioned), and then we can decode the challenge pretty easily without the need to append `0x` in front.

It also makes the string a bit more "self-describing" (i.e., by looking at it, we can tell that it is hex-encoded).